CC: Manual Work is a Bug (#1805) (#2089)
betterscientificsoftware/bssw.io: CuratedContent/ManualWorkIsABug.md

## Manual Work is a Bug

<!-- deck text start -->
Work to automate routine computing processes using an Agile incremental approach, starting by documenting the manual steps and then incrementally automating pieces of the process using scripts/code.
Continuing to do routine processes repeatedly with manual execution is a bug!
<!-- deck text end -->

#### Contributed by [Roscoe A. Bartlett](https://github.com/bartlettroscoe "Roscoe A. Bartlett")
#### Publication date: May 27, 2024

Resource information | Details
:--- | :---
Paper Title | Manual Work is a Bug
Authors | Thomas A. Limoncelli
Publication | [ACM Queue, March 14, 2018, Volume 16, Issue 1](https://queue.acm.org/detail.cfm?id=3197520)

Nearly everyone who works with computers often has to perform routine tasks.
Some people like system administrators (sysadmins) and DevOps engineers have **many** such tasks to perform.
However, even other types of developers and users often have repetitive tasks that they need to execute repeatedly.
And others in their teams and organizations often need to perform similar processes/tasks (or the same tasks, with minor variations).

> These routine processes may take long enough, be complex enough, have severe enough consequences if they are done in correctly, and/or be likely to be repeated enough to justify automating them to some degree.

Reviewer: ```suggestion
These routine processes may be time-consuming, be complex, have severe consequences if they are done incorrectly, and/or be likely to be repeated enough to justify automating them to some degree.
```

Author: I was trying to emphasize that there my be thresholds for time, complexity, severity, and/or future repeatability to trigger considering automating the process.   Does the updated text make that clear?

Author: Applied in commit 34a11ccbd84db1ea44bcf3eaea9c31c584cfe4fa (except I removed "be" from "be complex").

Thomas A. Limoncelli's article "Manual Work is a Bug" describes an Agile incremental process for developing increasing levels of automation for routine processes like these.
(Thomas mainly speaks to sysadmins in this article, but the process and the principles described are the same for any individual or team that needs to perform routine tasks with computers.)

> The basic idea is to start by writing documentation right from the beginning while manually performing the steps for the first time for every process that is a potential candidate to become a routine (automatable) process.

Reviewer: ```suggestion
The basic idea is to start by writing documentation right from the beginning while manually performing the steps, for the first time, for every process that is a potential candidate to become a routine (automated) process.
```

Author: Applied in commit 34a11ccbd84db1ea44bcf3eaea9c31c584cfe4fa.

The end state is a largely automated process that will save developers' time and avoid mistakes.

> This culture can be summarized in two sentences:
> (1) Every manual action must have a dual purpose of completing a task and improving the system.
> (2) Manual work should not be tolerated unless it generates an artifact or improves an existing one.

This Agile automation process is broken down into four phases:

* **Phase 1: Document the steps**: Document the process as you do the steps manually the first time.

* **Phase 2: Create automation equivalents**: When executing the process again, go back over the process manually following the existing documentation while refining the process and adding exact command-line snippets to the documentation.

* **Phase 3: Create automation**: Develop an expanding set of scripts/code to automate parts of the process as it is executed repeatedly.
(Every process iteration should expand the automation scripts/code scope and reduce the manual effort.)

* **Phase 4: Self-service autonomous systems**:  The final phase is to turn the process into a stand-alone tool using the developed scripts/code that runs with a single invocation.

> In some cases, such processes justify being set up as an ***autonomous process** (i.e., a process that runs automatically without any explicit action by the user.)

Reviewer: ```suggestion
In some cases, such processes justify being set up as an **autonomous process** (i.e., a process that runs automatically without any explicit action by the user.)
```

Author: Applied in commit 34a11ccbd84db1ea44bcf3eaea9c31c584cfe4fa.

Some people like system administrators (sysadmins) and DevOps engineers have **many** such tasks to perform.
However, even other types of developers and users often have repetitive tasks that they need to execute repeatedly.
And others in their teams and organizations often need to perform similar processes/tasks (or the same tasks, with minor variations).

> These routine processes may be time-consuming, complex, have severe consequences if they are done incorrectly, and/or be likely to be repeated enough to justify automating them to some degree.

Reviewer: ```suggestion
These routine processes may be time-consuming or complex, have severe consequences if they are done incorrectly, and/or be likely to be repeated enough to justify automating them to some degree.
```
